Buttplug.Core/ButtplugDevice.cs

```diff
@@ -21,30 +23,89 @@ public bool IsConnected
             }
         }
 
+        public uint VibratorCount
+        {
+            get
+            {
+                return _vibratorCount;
+            }
+
+            protected set
+            {
+                var tmp = new double[value];
+                for (var i = 0; i < value; i++)
+                {
+                    tmp[i] = i < _vibratorCount ? _vibratorSpeeds[i] : 0;
+                }
+
+                _vibratorCount = value;
+                _vibratorSpeeds = tmp;
+            }
+        }
+
         [CanBeNull]
         public event EventHandler DeviceRemoved;
 
+        [CanBeNull]
+        public event EventHandler<MessageReceivedEventArgs> MessageEmitted;
+
         [NotNull]
         protected readonly IButtplugLog BpLogger;
+
+        [NotNull]
+        protected readonly Dictionary<Type, ButtplugDeviceWrapper> MsgFuncs;
+
         [NotNull]
-        protected readonly Dictionary<Type, Func<ButtplugDeviceMessage, Task<ButtplugMessage>>> MsgFuncs;
+        protected double[] _vibratorSpeeds;
+
         private bool _isDisconnected;
 
+        private uint _vibratorCount;
+
+        public class ButtplugDeviceWrapper
+        {
+            public Func<ButtplugDeviceMessage, Task<ButtplugMessage>> Func;
+            public MessageAttributes Attrs;
+
+            public ButtplugDeviceWrapper(Func<ButtplugDeviceMessage, Task<ButtplugMessage>> aFunc, MessageAttributes aAttrs = null)
+            {
+                Func = aFunc;
+                Attrs = aAttrs;

```

Comment:
You can turn the assignment and if block into a single line.

Attrs = aAttrs ?? new MessageAttributes();